fix pg not deleted after pod created by deployment is deleted

lucming opened the pull request:
#https://github.com/volcano-sh/volcano/issues/1853

lucming commented:
/assign @shinytang6

shinytang6 commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -49,6 +49,24 @@ func (pg *pgcontroller) addPod(obj interface{}) {
+		return
+	}
+}
+
```
i'm ok about the changes, but is there a way to filter out the pods under vcjob? Because currently all pods are watched, but when the pods under vcjob are deleted, deleting pg will throw an error

Thor-wl replied:
Agree with @shinytang6 .Without the filter, just delete the pg is not so reasonable when some pod is deleted.

shinytang6 reviewed (changes requested):
i'm ok about the changes, but is there a way to filter out the pods under cjob? Because currently all pods are watched, but when the pods under vcjob are deleted, deleting pg will throw an error

Thor-wl commented:
/hold

hzxuzhonghu commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -76,7 +76,8 @@ func (pg *pgcontroller) Initialize(opt *framework.ControllerOption) error {
 	pg.podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: pg.addPod,
+		AddFunc:    pg.addPod,
+		DeleteFunc: pg.deletePG,
```
This is not good to call some time costing logic in eventhandler.

lucming replied:
Perhaps i can wirite the deleted pod in a queue,just like create pg

lucming replied:
> 

Perhaps i can wirite the deleted pod in a queue,just like create pg,do you think that's ok

hzxuzhonghu commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -49,6 +49,24 @@ func (pg *pgcontroller) addPod(obj interface{}) {
+		return
+	}
+
+	if err := pg.vcClient.SchedulingV1beta1().PodGroups(pod.Namespace).Delete(context.TODO(), pgName, metav1.DeleteOptions{}); err != nil {
```
does it mean delete the pg once a pod deleted?

lucming replied:
yes

shinytang6 commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -139,6 +142,8 @@ func (pg *pgcontroller) processNextReq() bool {
 	// If no error, forget it.
 	pg.queue.Forget(req)
 
+	pg.deletePG()
```
it's better to put this into another new goroutine

shinytang6 commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -139,6 +142,8 @@ func (pg *pgcontroller) processNextReq() bool {
 	// If no error, forget it.
 	pg.queue.Forget(req)
 
+	go pg.deletePG()
```
l think you misunderstood what l mean, l mean it's better to use another worker to delete pg, while not doing it in `processNextReq `

lucming replied:
so,you hopes i use first version that delete pg in delete eventhandler,but use a gorutine to do it,is that so?

shinytang6 replied:
@lucming No, l mean you can increase a `deleteWorker` goroutine to do that.
Something like that:
```
go wait.Until(pg.deleteWorker, 0, stopCh)

func (pg *pgcontroller) deleteWorker() {
	for pg.processNextDeleteReq() {}
}

func (pg *pgcontroller) processNextDeleteReq() {
        ...
}
```

lucming replied:
@shinytang6 done

Thor-wl commented:
/hold cancel

shinytang6 commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -94,6 +97,7 @@ func (pg *pgcontroller) Run(stopCh <-chan struct{}) {
 	cache.WaitForCacheSync(stopCh, pg.podSynced, pg.pgSynced)
 
 	go wait.Until(pg.worker, 0, stopCh)
+	go wait.Until(pg.deletePG, 0, stopCh)
```
you need a for-loop here, just like the `worker()` :)
```
func (pg *pgcontroller) worker() {
	for pg.processNextReq() {
	}
}
```

lucming replied:
@shinytang6 done

shinytang6 commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -17,6 +17,8 @@ limitations under the License.
 
 import (
+	"context"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
pls add one blank line between external and sys package

shinytang6 commented:
/ok-to-test

Thor-wl commented:
@lucming Pls rebase all the commits into one.

shinytang6 commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -150,3 +157,32 @@ func contains(slice []string, element string) bool {
+	defer pg.pgDeletedQueue.Done(req)
+
+	if _, err := pg.pgLister.PodGroups(req.pgNamespace).Get(req.pgName); err != nil {
+		return false
```
l think it should not return false here

lucming replied:
@Thor-wl done

shinytang6 commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -150,3 +157,32 @@ func contains(slice []string, element string) bool {
+
+	if err := pg.vcClient.SchedulingV1beta1().PodGroups(req.pgNamespace).Delete(context.TODO(), req.pgName, metav1.DeleteOptions{}); err != nil {
+		klog.Errorf("Failed to delete volcano job,ns:%s,name:%s,err:%v", req.pgNamespace, req.pgName, err)
+		return false
```
ditto

shinytang6 reviewed (commented):
Thx, generally lgtm now 👍 
/cc @Thor-wl @hwdef @william-wang pls take another look

hwdef commented on pkg/controllers/podgroup/pg_controller.go:
```diff
@@ -150,3 +157,32 @@ func contains(slice []string, element string) bool {
+	req := obj.(pgRequest)
+	defer pg.pgDeletedQueue.Done(req)
+
+	if _, err := pg.pgLister.PodGroups(req.pgNamespace).Get(req.pgName); err != nil {
```
Did we ignore this error?

hwdef commented:
/ok-to-test

k82cn commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -49,6 +55,44 @@ func (pg *pgcontroller) addPod(obj interface{}) {
 	pg.queue.Add(req)
 }
 
+func (pg *pgcontroller) deletePod(obj interface{}) {
```
Are we going to delete `PodGroup` when one Pod deleted?

lucming replied:
> IMO, we should handle this case by GC, e.g. only delete PodGroup when its owner was deleted.

I agree with you,the essence of this problem is wrong to set the owner of pod,but it's a little hard to find the outest controller by pod,there are too many upstream pods that can be created,and we cannot enumerate explicit outputs.

k82cn replied:
> > IMO, we should handle this case by GC, e.g. only delete PodGroup when its owner was deleted.
> 
> I agree with you,the essence of this problem is wrong to set the owner of pod,but it's a little hard to find the outest controller by pod,there are too many upstream pods that can be created,and we cannot enumerate explicit outputs.

We can find controlled owner reference of Pod to do that :)

lucming replied:
ok, let me change that

k82cn commented:
IMO, we should handle this case by GC, e.g. only delete PodGroup when its owner was deleted.

k82cn commented:
/hold

k82cn commented on pkg/controllers/podgroup/pg_controller_handler.go:
```diff
@@ -125,10 +126,14 @@ func (pg *pgcontroller) createNormalPodPGIfNotExist(pod *v1.Pod) error {
+			if ownerReference.Controller != nil &&
+				*ownerReference.Controller &&
+				ownerReference.APIVersion == batchv1alpha1.SchemeGroupVersion.String() &&
+				ownerReference.Kind == "Job" {
```
hm... how this PR fix the problem?

lucming replied:
Only pods created by the controller of volcano use the owner of the pod as the owner of the pg,in other case,pg will use pod as the owner of pg.

Thor-wl commented:
Any update for this PR? Seems some problems in UT. @lucming

lucming commented:
> Any update for this PR? Seems some problems in UT. @lucming

i need modify this unit-test,in this case that pod created by rs,we should use pod as pg's owner,because the lifetime of pg may be different with rs,rs maybe craeted by deployment,use pod as pg's owner maybe the probably direct way,and what do you think?

Thor-wl commented:
> > Any update for this PR? Seems some problems in UT. @lucming
> 
> i need modify this unit-test,in this case that pod created by rs,we should use pod as pg's owner,because the lifetime of pg may be different with rs,rs maybe craeted by deployment,use pod as pg's owner maybe the probably direct way,and what do you thin

> > Any update for this PR? Seems some problems in UT. @lucming
> 
> i need modify this unit-test,in this case that pod created by rs,we should use pod as pg's owner,because the lifetime of pg may be different with rs,rs maybe craeted by deployment,use pod as pg's owner maybe the probably direct way,and what do you think?

Well, I don't think set the pg's owner reference to be pod is reasonable. In fact, we should create pg for deployment, statefulset, job, vcjob or some other resource types. If it's pod alone, I think it's ok to create pg for it.

lucming commented:
> > > Any update for this PR? Seems some problems in UT. @lucming
> > 
> > 
> > i need modify this unit-test,in this case that pod created by rs,we should use pod as pg's owner,because the lifetime of pg may be different with rs,rs maybe craeted by deployment,use pod as pg's owner maybe the probably direct way,and what do you thin
> 
> > > Any update for this PR? Seems some problems in UT. @lucming
> > 
> > 
> > i need modify this unit-test,in this case that pod created by rs,we should use pod as pg's owner,because the lifetime of pg may be different with rs,rs maybe craeted by deployment,use pod as pg's owner maybe the probably direct way,and what do you think?
> 
> Well, I don't think set the pg's owner reference to be pod is reasonable. In fact, we should create pg for deployment, statefulset, job, vcjob or some other resource types. If it's pod alone, I think it's ok to create pg for it.

we can get resource easy from deloyment->rs->pod,but it's hard to go from bottom to top,eg:pod->rs->deployment,maybe i can try

Thor-wl commented:
/cc @william-wang @shinytang6 Do you have some good ideas?

lucming commented:
`func getOutestOwner(dyClient dynamic.Interface, kubeClient kubernetes.Interface, object *unstructured.Unstructured, result *[]metav1.OwnerReference) {
	owners := object.GetOwnerReferences()
	if len(owners) > 0 {
		for _, owner := range owners {
			if owner.Controller != nil {
				apiversionList := strings.Split(owner.APIVersion, "/")
				var group, version string
				if len(apiversionList) == 2 {
					group, version = apiversionList[0], apiversionList[1]
				} else if len(apiversionList) == 1 {
					version = apiversionList[0]
				}
				gk := schema.GroupKind{
					Group: group,
					Kind:  owner.Kind,
				}

				groupResources, err := restmapper.GetAPIGroupResources(kubeClient.Discovery())
				if err != nil {
					klog.Errorf("failed to get resouces(%s:%s)in ns(%s),error=%v", gk.String(), owner.Name, object.GetNamespace())
					return
				}
				rm := restmapper.NewDiscoveryRESTMapper(groupResources)
				mapping, err := rm.RESTMapping(gk, version)
				parentObj, err := dyClient.Resource(mapping.Resource).
					Namespace(object.GetNamespace()).
					Get(context.Background(), owner.Name, metav1.GetOptions{})
				if err != nil {
					klog.Errorf("failed to get resouces(%s:%s)in ns(%s),error=%v", gk.String(), owner.Name, object.GetNamespace())
					return
				}
				getOutestOwner(dyClient, kubeClient, parentObj, result)
			}
		}
	} else {
		ref := metav1.NewControllerRef(object, object.GroupVersionKind())
		*result = append(*result, *ref)
	}
}`

lucming commented:
```
func getOutestOwner(dyClient dynamic.Interface, kubeClient kubernetes.Interface, object *unstructured.Unstructured, result *[]metav1.OwnerReference) {
	owners := object.GetOwnerReferences()
	if len(owners) > 0 {
		for _, owner := range owners {
			if owner.Controller != nil {
				apiversionList := strings.Split(owner.APIVersion, "/")
				var group, version string
				if len(apiversionList) == 2 {
					group, version = apiversionList[0], apiversionList[1]
				} else if len(apiversionList) == 1 {
					version = apiversionList[0]
				}
				gk := schema.GroupKind{
					Group: group,
					Kind:  owner.Kind,
				}

				groupResources, err := restmapper.GetAPIGroupResources(kubeClient.Discovery())
				if err != nil {
					klog.Errorf("failed to get resouces(%s:%s)in ns(%s),error=%v", gk.String(), owner.Name, object.GetNamespace())
					return
				}
				rm := restmapper.NewDiscoveryRESTMapper(groupResources)
				mapping, err := rm.RESTMapping(gk, version)
				parentObj, err := dyClient.Resource(mapping.Resource).
					Namespace(object.GetNamespace()).
					Get(context.Background(), owner.Name, metav1.GetOptions{})
				if err != nil {
					klog.Errorf("failed to get resouces(%s:%s)in ns(%s),error=%v", gk.String(), owner.Name, object.GetNamespace())
					return
				}
				getOutestOwner(dyClient, kubeClient, parentObj, result)
			}
		}
	} else {
		ref := metav1.NewControllerRef(object, object.GroupVersionKind())
		*result = append(*result, *ref)
	}
}
```
I've implemented a way to find the outermost controller recursively, so you can look at it and give me some feedback
